main/tests/server/src/com/google/refine/tests/commands/project/SetProjectMetadataCommandTests.java

```diff
@@ -0,0 +1,212 @@
+        
+        ProjectMetadata metadata = new ProjectMetadata();
+        metadata.setUserMetadata(new JSONArray("[ {name: clientID, display: true} ]"));
+        
+        // mock dependencies
+        when(request.getParameter("project")).thenReturn(PROJECT_ID);
+        when(projMan.getProject(anyLong())).thenReturn(proj);
+        when(proj.getMetadata()).thenReturn(metadata);
+        
+        try {
+            when(response.getWriter()).thenReturn(pw);
+        } catch (IOException e1) {
+            Assert.fail();
+        }
+    }
+
+    @AfterMethod
+    public void TearDown() {
+        SUT = null;
+
+        projMan = null;
+        ProjectManager.singleton = null;
+        proj = null;
+        pw = null;
+        request = null;
+        response = null;
+    }
+
+    /**
+     *  Contract for a complete working post
+     */
+    @Test
+    public void setMetadataTest() {
+        when(request.getParameter("name")).thenReturn("subject");
+        when(request.getParameter("value")).thenReturn(SUBJECT);
+        
+        // run
+        try {
+            SUT.doPost(request, response);
+        } catch (ServletException e) {
+            Assert.fail();
+        } catch (IOException e) {
+            Assert.fail();
+        }
+
+        // verify
+        verify(request, times(2)).getParameter("project");      
+        verify(projMan, times(1)).getProject(PROJECT_ID_LONG);
+
+        verify(response, times(1))
+                .setHeader("Content-Type", "application/json");
+        verify(proj, times(1)).getMetadata();
+        try {
+            verify(response, times(1)).getWriter();
+        } catch (IOException e) {
+            Assert.fail();
+        }
+        verify(pw, times(1)).write("{ \"code\" : \"ok\" }");
+        
+        Assert.assertEquals(proj.getMetadata().getSubject(), SUBJECT);
+    }
+    
+    /**
+     *  set a user defined metadata field
+     * @throws JSONException 
+     */
+    @Test
+    public void setUserMetadataFieldTest() throws JSONException {
+        when(request.getParameter("name")).thenReturn("clientID");
+        when(request.getParameter("value")).thenReturn("IBM");
+        
+        // run
+        try {
+            SUT.doPost(request, response);
+        } catch (ServletException e) {
+            Assert.fail();
+        } catch (IOException e) {
+            Assert.fail();
+        }
+
+        // verify
+        verify(request, times(2)).getParameter("project");      
+        verify(projMan, times(1)).getProject(PROJECT_ID_LONG);
+
+        verify(response, times(1))
+                .setHeader("Content-Type", "application/json");
+        verify(proj, times(1)).getMetadata();
+        try {
+            verify(response, times(1)).getWriter();
+        } catch (IOException e) {
+            Assert.fail();
+        }
+        verify(pw, times(1)).write("{ \"code\" : \"ok\" }");
+        
+        JSONObject obj = (JSONObject) proj.getMetadata().getUserMetadata().get(0);
```

Comment:
hmm, wondering about maximums here for UserMetadata ... like if they use a client API for filling it instead ?
Coming from telecom, I'm always worried probably too much about maximums and seg faulting :)

Reply:
I believe there is max size of JSONObject. need to dig more on this.

Reply:
👍